src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs

```diff
@@ -545,15 +563,17 @@ public void SendMessages(IEnumerable<RequestMessage> messages, MessageEncoderSet
             Ensure.IsNotNull(messages, nameof(messages));
             ThrowIfDisposedOrNotOpen();
 
-            var helper = new SendMessagesHelper(this, messages, messageEncoderSettings);
+            var helper = new SendMessagesHelper(this, messages, _sendCompressorType, _compressorSource, messageEncoderSettings);
             try
             {
                 helper.EncodingMessages();
-                using (var buffer = helper.EncodeMessages(cancellationToken))
+                var buffers = helper.EncodeMessages(cancellationToken);
+                using (var possiblyCompressedBuffer = buffers.PossiblyCompressedBuffer)
+                using (var originalMessageBuffer = buffers.OriginalMessageBuffer)
                 {
-                    helper.SendingMessages(buffer);
-                    SendBuffer(buffer, cancellationToken);
-                    helper.SentMessages(buffer.Length);
+                    helper.SendingMessages(originalMessageBuffer);
+                    SendBuffer(possiblyCompressedBuffer, cancellationToken);
+                    helper.SentMessages(possiblyCompressedBuffer.Length);
```

Comment:
Let's refactor in a way that leaves the SendMessagesHelper untouched:

                using (var uncompressedBuffer = helper.EncodeMessages(cancellationToken))
                {
                    helper.SendingMessages(uncompressedBuffer);
                    int sentLength;
                    if (AnyMessageNeedsToBeCompressed(messages))
                    {
                        using (var compressedBuffer = CompressMessages(messages, uncompressedBuffer, messageEncoderSettings))
                        {
                            SendBuffer(compressedBuffer, cancellationToken);
                            sentLength = compressedBuffer.Length;
                        }
                    }
                    else
                    {
                        SendBuffer(uncompressedBuffer, cancellationToken);
                        sentLength = uncompressedBuffer.Length;
                    }
                    helper.SentMessages(sentLength);
                }

Part of the idea is that if no messages need to be compressed then we are essentially executing exactly the same code path as before, which is a smaller diff and should reduced the chances of bugs.

Another part of the idea is to properly put the buffers used in using statements.

Reply:
New helper method:

        private bool AnyMessageNeedsToBeCompressed(IEnumerable<RequestMessage> messages)
        {
            return _sendCompressorType.HasValue && messages.Any(m => m.MayBeCompressed);
        }

Reply:
New helper method:

        private IByteBuffer CompressMessages(
            IEnumerable<RequestMessage> messages,
            IByteBuffer uncompressedBuffer,
            MessageEncoderSettings messageEncoderSettings)
        {
            var outputBufferChunkSource = new OutputBufferChunkSource(BsonChunkPool.Default);
            var compressedBuffer = new MultiChunkBuffer(outputBufferChunkSource);

            using (var uncompressedStream = new ByteBufferStream(uncompressedBuffer, ownsBuffer: false))
            using (var compressedStream = new ByteBufferStream(compressedBuffer, ownsBuffer: false))
            {
                foreach (var message in messages)
                {
                    var uncompressedMessageLength = uncompressedStream.ReadInt32();
                    uncompressedStream.Position -= 4;

                    using (var uncompressedMessageSlice = uncompressedBuffer.GetSlice((int)uncompressedStream.Position, uncompressedMessageLength))
                    using (var uncompressedMessageStream = new ByteBufferStream(uncompressedMessageSlice, ownsBuffer: false))
                    {
                        if (message.MayBeCompressed)
                        {
                            CompressMessage(message, uncompressedMessageStream, compressedStream, messageEncoderSettings);
                        }
                        else
                        {
                            uncompressedMessageStream.EfficientCopyTo(compressedStream);
                        }
                    }
                }
            }

            return compressedBuffer;
        }

        private void CompressMessage(
            RequestMessage message,
            ByteBufferStream uncompressedMessageStream,
            ByteBufferStream compressedStream,
            MessageEncoderSettings messageEncoderSettings)
        {
            var compressedMessage = new CompressedMessage(message, uncompressedMessageStream, _sendCompressorType.Value);
            var compressedMessageEncoderFactory = new BinaryMessageEncoderFactory(compressedStream, messageEncoderSettings, _compressorSource);
            var compressedMessageEncoder = compressedMessageEncoderFactory.GetCompressedMessageEncoder(null);
            compressedMessageEncoder.WriteMessage(compressedMessage);
        }